GetTuples should handle sub-sparse matrix correctly

siddharth178 opened the pull request:
GetTuples could be called on a submatrix of bigger sparse matrix. It should handle offset and computed row and col indexes correctly so that it doesn't access other elements from parent sparse matrix and generate wrong row/col indexes for tuples.

Some more background of this code change - 

consider [last commit](https://github.com/clyphub/gillnet/commit/39cf7c39fd3fa61f6984e3acd243d78bb2abed9f) from  branch https://github.com/clyphub/gillnet/compare/main-32808-tocplex-noexpand

this is no-expand implementation of ToCPLEX. here we are trying to directly iterate over math constraints and use DecisionVars and ElasticVars sparse matrices to generate CPLEXConstraints.

while testing, we found that this implementation is generating constraints with x and y terms having -ve indexes in their name. see this snippet from lp_mode.lp
```
  c1: \ constraint 1
  \ constraint_id,flight_id,label,value,target_value,min_value,max_value,measurement,reference,is_inelastic
  \ 1,3,flight_id,3,50000,50000,50000,budget,,true
  + 250 x-5
  + 250 x-4
  + 1000 x-3
  + 1000 x-2
  + 1000 x-1
  + 250 x0
  + 250 x1
  + 250 x2
  + 1000 x3
  + 1000 x4
  + 1000 x5
  \ c1: 'constraint_id,flight_id,label,value,target_value,min_value,max_value,measurement,reference,is_inelastic
  \ 1,3,flight_id,3,50000,50000,50000,budget,,true
  \ ' has only zero y terms
  + 0 y0
    >= + 500
```

digging more into it we got to the implementation of `GetTuples` that we call from `writeSparseNonZeroTerms`. thats the problematic code. `GetTuples` implementation doesn't consider offset, x and y limits if a submatrix (or row/col vector) of bigger sparse matrix was used. thats how it starts returning slice of IndexedValues with negative (or out of range) row and col indexes in them.

till this time, because we were expanding matrix and we were creating a copy of xCoeffs and yCoeffs, each matrix was independent. so GetTuples worked fine. but now we are sharing decisionvars and elasticvars matrices so it started failing.

Ref: https://clypdinc.atlassian.net/browse/MAIN-32808

Commit:
GetTuples should handle sub-sparse matrix correctly

GetTuples could be called on a submatrix of bigger sparse matrix. It
should handle offset and computed row and col indexes correctly so that
it doesn't access other elements in parent sparse matrix and generate
wrong row/col indexes for tuples

joey-clypd commented on sparse_test.go:
```diff
@@ -202,6 +202,43 @@ func TestGetTuples_Sparse(t *testing.T) {
 	}
+
+	B := MakeDenseMatrix([]float64{0, 0, 0, 40, 0, 60, 0, 80, 90}, 3, 3).SparseMatrix()
+	B0 := B.GetRowVector(0)
```
Do you mind using SubMatrix for this and testing the column bounds as well?

joey-clypd reviewed (changes requested):
One minor change.

ashultz0 commented:
I'm having trouble reviewing because there's no description of what the problem was which would usually be in a jira ticket. What was the pre-fix behavior?

I certainly have no problem believing there were horrible bugs in this library, having interacted with it before.

siddharth178 commented:
@ashultz0 I have updated description with some background to this change.

APILLSBURY reviewed (commented):
Took a look but I don't understand enough about the matrix code to be able to effectively review. If you both (@joey-clypd  and @siddharth178 ) agree that this is correct though I think you're fine to merge it.

Commit:
fix GetRowColIndex

this now panics if invalid index is provided as input

Commit:
fix GetColVector

interestingly, we don't use it in any of our code. so we are safe.

Commit:
cleanup GetTuples

it no more accepts row as input and returns all the non-zero values from
the matrix

the way we are using it in gillnet code is always with 1 row, so no
changes needed there

it now uses correct/better version of GetRowColIndex

Commit:
remove debug print line

joey-clypd commented on sparse_test.go:
```diff
@@ -234,57 +234,74 @@ func Test_GetRowColIndex(t *testing.T) {
 
 func TestGetTuples_Sparse(t *testing.T) {
 	A := NormalsSparse(4, 4, 16)
+	rowVals := []float64{}
```
Hooray for tests!
